### ISSUE TITLE
Sketchup.temp_dir returns short names on windows

### ISSUE THREAD

**Issue by tdapper** [SketchUp/api-issue-tracker#135]
When using `Sketchup.temp_dir` on Windows and the path has a folder component that does not comply with the 8.3 naming scheme, it returns the short name (e.g. `C:\Users\MyLongUsername\AppData\Local\Temp` becomes `C:\Users\MYLONG~1\AppData\Local\Temp`).

This can cause all sorts of problems e.g. when comparing this path with paths from other sources that don't have this problem.

The problem might come from the underlying integration of Ruby or general use of the Win32 API, because using regular Ruby functions such as `Dir.mktmpdir()` produce the same problem.

SketchUp 2018
Windows

**Comment by MSP-Greg** [comment]
This is a messy issue.  Is it Windows, Ruby, or SU's implementation of Ruby.

I added a 'long user name' account on Win10, and `ENV['TEMP']` and `ENV['TMP']` were set by Windows to 8.3 naming.  Ruby's `Etc.systmpdir` did correctly return the long name.

I haven't checked all versions, but newer Ruby versions check `ENV['TMPDIR']` before the other ENV settings.  You might experiment with setting that to the long name when your code loads...

**Comment by tdapper** [comment]
You are right, it certainly does look a little messy. I checked the environment as suggested and it indeed already contains the short names. The temporary folder detection in Python appears to do the same thing (return the short path). 

This makes it look like it's not a SketchUp issue. One could perhaps argue, that it would be cleaner if Ruby would call the Win32 `GetLongPathName()` function internally, but then again there are probably also reasons against that (even though it looks like that is what `Etc.systmpdir` does).

Not sure what to say other than that I would kinda implicitly expect the Sketchup API function not to bother with legacy 8.3 file name conventions. That is also what @thomthom appears to expect when he suggested to use it. So I'd tend to lean towards changing that behavior on the SketchUp-side from a user expectation point of view. It might break stuff in some rare cases, though.

**Comment by taustin73** [comment]
Logged SU-41959

**Comment by tdapper** [comment]
We're looking into this again currently, because we are working around this by using `Win32API`, which now produces deprecation messages. I was wondering whether this has been solved somewhere for newer versions of SketchUp (specifically SketchUp 2021, which now uses Ruby 2.7), @taustin73, @thomthom, @MSP-Greg.

**Comment by thomthom** [comment]
AFIK there's been no changes in SketchUp itself. And I don't have a machine with a long username available right at this moment.

As for `Win32API` deprecation, `Fiddle` is a cross platform successor. This ships with Ruby standard library.

**Comment by tdapper** [comment]
@thomthom Thanks for your reply. I am aware of `fiddle`, but since what we did with it was a workaround to a problems with `Sketchup.temp_dir` as well as `Dir.mktmpdir` in the first place, I was hoping that this may have been fixed (probably on the Ruby side) and I could perhaps just use the functions as intended and wouldn't have to port that workaround at all.

**Comment by MSP-Greg** [comment]
Ruby bug report:
https://bugs.ruby-lang.org/issues/15247, no responses.  I suppose I could post another message.

Part of the problem with this issue is that fixes would be a breaking change, and I suspect that most users of '*nix' based software are aware of issues with file systems.  Not necessarily true of Windows SU users.

I just looked thru the Etc c source, and it appears that Ruby doesn't really look up the user's `ENV['temp']` setting and translate it to a long name.  It looks up `ENV['LOCALAPPDATA']`, which is stored as a long name, and appends `Temp` to it.

Have you considered using `File.absolute_path`?  I think it does the short/long path conversion...

**Comment by DanRathbun** [comment]
You should be able to use the [`WIN32OLE`](https://ruby-doc.org/stdlib-2.5.5/libdoc/win32ole/rdoc/index.html) class.

```ruby
if Sketchup.platform == :platform_win
  require "win32ole"
  def get_temp_path
    WIN32OLE.new("WScript.Shell").ExpandEnvironmentStrings("%TEMP%")
  end
else
  def get_temp_path
    ENV["TEMP"] || ENV["TMP"]
  end
end
```

REF: Windows Scripting Host object

[`WshShell Object`](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/windows-scripting/aew9yb99(v=vs.84))
[`ExpandEnvironmentStrings Method`](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/windows-scripting/dy8116cf(v=vs.84))

**Comment by MSP-Greg** [comment]
@DanRathbun

That doesn't work if `ENV['TEMP'] is a short filename.

**Comment by DanRathbun** [comment]
Bummer. I thought it would expand the `User~1`, but I guess it means that it expands the `%name%` variable.

----

So apparently even Windows has left the `TEMP` with the shortened names for backward compatibility ?

```ruby
TEMP = ENV["TEMP"]
if TEMP.include?('~') &&
TEMP.start_with?("#{ENV['SYSTEMDRIVE']}\\Users\\") &&
TEMP.end_with?('\AppData\Local\Temp')
  TEMP = File.join(ENV['LocalAppData'],'Temp')
end
```

**Comment by MSP-Greg** [comment]
> Windows has left the TEMP with the shortened names for backward compatibility ?

Appears so.

All the recent code testing I did was with stand-alone Ruby master/main.

I just checked with SU 2018 (Ruby 2.2.4), and `File.absolute_path ENV['TEMP']` correctly translates a short path to a long path.

Also, checked with SU 2018 & 2021, and `Sketchup.temp_dir` is static in the sense that changing `ENV['Temp']` after SU is running doesn't change it.

I don't have any older versions loaded...

**Comment by thomthom** [comment]
> I was hoping that this may have been fixed (probably on the Ruby side) and I could perhaps just use the functions as intended and wouldn't have to port that workaround at all.

It looks like `GetTempPath` is always returning 8.3 for app compatibility, and that using `GetLongPathName` on that is the canonical way to go. We could do that for `Sketchup.temp_dir`, but I'm also concerned about compatibility in that it might break existing functions that already use it. But we could make it opt-in to get the long name; `Sketchup.temp_dir(longname: true)` (or something to that effect.)

**Comment by DanRathbun** [comment]
> ... but I'm also concerned about compatibility in that it might break existing functions that already use it. But we could make it opt-in to get the long name; `Sketchup.temp_dir(longname: true)` (or something to that effect.)

That would be the best backward compatible way as older versions would ignore the named argument.)

----

**However**, this issue only occurs because this method should never have been implemented within the **SketchUp API**.

It does not belong in the SketchUp API. It belongs in a System library (or gem) like those pointed at by Greg. In this way it can be fixed and updated independently from SketchUp.

If gems are going to be a problem going forward, then a community maintained library would be the best option in my opinion.

**Comment by thomthom** [comment]
Yea, I agree, it wasn't an ideal case for the SketchUp API. Unfortunately there wasn't any good alternative at the time. (Nor at this time for that matter.)

**Comment by DanRathbun** [comment]
What about SketchUp updating it's **own** distribution of the **[`Etc`](https://ruby-doc.org/stdlib-2.7.1/libdoc/etc/rdoc/index.html)** libarary ? (This could be a community Git effort.)

**Comment by thomthom** [comment]
We ship what the Ruby project provide. We don't want to be patching Ruby as that causes more work when we update. If anything I'd recommend anyone interested providing a PR to the upstream library.

**Comment by DanRathbun** [comment]
So we are left with a workaround such as ...
```ruby
# At the top of an extension submodule or
# ... within some library module (perhaps a mixin):

LOCALAPPDATA ||= File.absolute_path(ENV['LOCALAPPDATA'])
TEMP ||= File.join(LOCALAPPDATA, File.basename(ENV['TEMP']))
```

**Comment by thomthom** [comment]
For the record here is (one of) the issues on this topic in the Ruby bug tracker: https://bugs.ruby-lang.org/issues/12650

**Comment by kalpana-ghodeswar** [comment]
Hi [tdapper](https://github.com/tdapper),

I just tried to reproduce this issue with version SU2023.0 onwards on Windows, and it's not returning short format any more. The issue has been resolved. Could you please confirm the same? thanks 

Output : 
`SketchUp Temp Dir: C:/Users/path/AppData/Local/Temp
Ruby Temp Dir: C:/Users/path/AppData/Local/Temp/d20250519-27344-oarbl6`
